live-build/misc/upgrade-scripts/upgrade-container

```diff
@@ -273,12 +273,149 @@ function run() {
+	# information, since we want to install the bootloader directly
+	# to the device, rather than to a partition of the device.
+	#
+	zpool list -vH rpool |
+		awk '! /rpool|mirror|replacing|spare/ {print $1}' |
+		sed 's/[0-9]*$//'
+}
+
+function migrate_password_for_user() {
+	local user="$1"
+	local password
+
+	password="$(awk -F: "\$1 == \"$user\" {print \$2}" /etc/shadow)"
+	chroot "/var/lib/machines/$CONTAINER" usermod -p "$password" "$user" ||
+		die "'usermod -p ... $user' failed for '$CONTAINER'"
+}
+
+function migrate_file() {
+	local path="$1"
+	local directory
+
+	if [[ -f "$path" ]]; then
+		directory="$(dirname "$path")"
+		mkdir -p "/var/lib/machines/${CONTAINER}${directory}" ||
+			die "'mkdir -p $directory' failed for '$CONTAINER'"
+		cp "$path" "/var/lib/machines/${CONTAINER}${path}" ||
+			die "'cp $path' failed for '$CONTAINER'"
+	fi
+}
+
+function convert_to_bootfs_cleanup() {
+	for dir in /proc /sys /dev; do
+		umount -R "/var/lib/machines/${CONTAINER}${dir}" ||
+			warn "'umount -R' of '$dir' failed"
+	done
+}
+
+#
+# The purpose of this function is to convert an existing upgrade
+# container (specified by the $CONTAINER global variable) into the next
+# boot filesystem used by the appliance; i.e. after calling this
+# function, the container's root filesystem will be used as the
+# appliance's root filesystem, the next time the appliance boots.
+#
+# This is done by updating the appliance's bootloader (i.e. grub) to
+# point to the container's filesystem, along with setting the mountpoint
+# of the filesystem to be "/" instead of "/var/lib/machines/$CONTAINER".
+# This effectively removes the container, since systemd-nspawn only
+# looks in /var/lib/machines, so it'll no longer find the container
+# after the mountpoint changes.
+#
+function convert_to_bootfs() {
+	trap convert_to_bootfs_cleanup EXIT
+
+	for dir in /proc /sys /dev; do
+		mount --rbind "$dir" "/var/lib/machines/${CONTAINER}${dir}" ||
+			die "'mount --rbind' of '$dir' failed"
+		mount --make-rslave "/var/lib/machines/${CONTAINER}${dir}" ||
+			die "'mount --make-rslave' of '$dir' failed"
+	done
+
+	#
+	# When performing a not-in-place upgrade, the root and delphix
+	# users will not have any password by default. Thus, we need to
+	# explicitly configure the passwords on the new root filesystem.
+	# Here, we ensure the passwords for the root and delphix user on
+	# the new root filesystem match their current values.
+	#
+	migrate_password_for_user delphix
+	migrate_password_for_user root
+
+	#
+	# LX-72 Until we have a proper solution for migrating the
+	# configuration of a given appliance to the new root filesystem
+	# (i.e. any configuration that isn't supplied by a package), we
+	# resort to explicitly migrating some of this configuration
+	# here. The files listed here are dynamically generated and/or
+	# modified by appliance-build.
+	#
+	while read -r file; do
+		migrate_file "$file"
+	done <<-EOF
```

Comment:
Did you forget `/etc/netplan/50-cloud-init.yaml` by any chance? Without it networking will be unconfigured.

Reply:
No. I'm relying on the same mechanism we use on first boot, which is our cloud-init configuration files that we migrate.

As we talked, it's probably only working because of a cloud-init bug, but I wanted to be consistent with how we configure networking on the initial boot.

Once we implement configuration migration "properly", we'll probably copy that file.. or w/e file the virtualization engine creates when configuring the network.